src/Stratis.Bitcoin/P2P/PeerConnector.cs

```diff
@@ -0,0 +1,219 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Threading;
+using System.Threading.Tasks;
+using NBitcoin;
+using NBitcoin.Protocol;
+using Stratis.Bitcoin.Utilities;
+
+namespace Stratis.Bitcoin.P2P
+{
+    public sealed class WellKnownGroupSelectors
+    {
+        private static Func<IPEndPoint, byte[]> byEndpoint;
+        public static Func<IPEndPoint, byte[]> ByEndpoint
+        {
+            get
+            {
+                return byEndpoint = byEndpoint ?? new Func<IPEndPoint, byte[]>((endpoint) =>
+                {
+                    var bytes = endpoint.Address.GetAddressBytes();
+                    var port = Utils.ToBytes((uint)endpoint.Port, true);
+                    var result = new byte[bytes.Length + port.Length];
+                    Array.Copy(bytes, result, bytes.Length);
+                    Array.Copy(port, 0, result, bytes.Length, port.Length);
+                    return bytes;
+                });
+            }
+        }
+
+        private static Func<IPEndPoint, byte[]> byNetwork;
+        public static Func<IPEndPoint, byte[]> ByNetwork
+        {
+            get
+            {
+                return byNetwork = byNetwork ?? new Func<IPEndPoint, byte[]>((ip) =>
+                {
+                    return IpExtensions.GetGroup(ip.Address);
+                });
+            }
+        }
+    }
+
+    public sealed class RelatedNodeGroups : Dictionary<string, PeerConnector>
+    {
+        public void Register(string name, PeerConnector nodeGroup)
+        {
+            if (nodeGroup != null)
+            {
+                this.Add(name, nodeGroup);
+                nodeGroup.RelatedGroups = this;
+            }
+        }
+
+        public IPEndPoint[] GlobalConnectedNodes()
+        {
+            IPEndPoint[] all = new IPEndPoint[0];
+            foreach (var kv in this)
+            {
+                var endPoints = kv.Value.ConnectedNodes.Select(n => n.RemoteSocketEndpoint).ToArray<IPEndPoint>();
+                all = all.Union<IPEndPoint>(endPoints).ToArray<IPEndPoint>();
+            }
+
+            return all;
+        }
+    }
+
+    public sealed class PeerConnector : IDisposable
+    {
+        internal PeerConnector(Network network, INodeLifetime nodeLifeTime, NodeConnectionParameters parameters, NodeRequirement nodeRequirements, Func<IPEndPoint, byte[]> groupSelector, IAsyncLoopFactory asyncLoopFactory)
+        {
+            this.disconnect = CancellationTokenSource.CreateLinkedTokenSource(nodeLifeTime.ApplicationStopping);
```

Comment:
Is there a reason we need a linked cancellation?
We will never just cancel `this.disconnect` on its own right?

Reply:
You are right, I have set ```this.disconnect``` to ```nodeLifetime.Application``` token in the constructor.

Reply:
I would suggest for consistency with other places just keep an instance of nodeLifetime and use it around (instead of disconnect )

Reply:
Ok will do!